### DISCUSSION TITLE
POLICYD: Add Cluebringer Support

### DISCUSSION THREAD

**Discussion by SomeGeek** [modoboa/modoboa#575] [Ideas]
We should integrate Postfix [Cluebringer](https://www.policyd.org) support. It's nearly essential when it comes to mail security. Plays nice with amavisd-new as well.

**Comment by tonioo** [comment]
Integrating modoboa with a policy server for Postfix is a good idea. I don't know if this one is a good choice since it doesn't seem to be maintained anymore...

Do you know another tool ?

**Comment by piwats** [comment]
Sqlgrey for greylisting, allows cluster mode db access, which comes handy with a multi mx setup.
Dnsbl can be done via postfix,
Some policy can be very good achieved by postfixs` own postcreen.

there is a good web interface for sqlgrey: greyface.
policyd-weight?

**Comment by carragom** [comment]
Policyd v1 is no longer maintained, but AFAIK policyd v2 AKA cluebringer is live and kicking. Latest stable release v2.014 was 2013-10-26. Is it stated somewhere that cluebringer is no longer maintained ?

**Comment by tonioo** [comment]
Indeed, it seems the project is still active. (looking at their gitlab)

What kind of intgration would you see ?

**Comment by SomeGeek** [comment]
Well, I should have stated more clearly that we should support their feautures.
- Access Control
- CheckHELO
- CheckSPF
- Greylisting (Per Domain)

For more info, check their website. Security is becoming very important these days, so such things should be possible to manage within Modoboa.

**Comment by piwats** [comment]
Is cluebringer realy needed? 
I see the problem, that cluebringer is not shipped with every distro today.

Then I see, that, if Access Control, CheckHELO, CheckSPF and Greylisting is demanded, I have to ask myself why do one need a separated daemon to run, if it is all feasible with already avaible postfix/amavisd/spamassassin/{grey|sql}list?

Please do not understand me wrong, but as far as I can see, all those features aforementioned are already avaible to the standard modoboa setup apart from greylisting.

I highly recomend to use sqlgrey, as it allows to be deployed in a cluster environment (but not limited to) with the ability to split read write. Also there is a nice webgui I highly recommend to everyone: 

https://github.com/teqneers/Greyface

It allows per IP/emailaddr./domain settings.

Access Control is also feasible with postfix as well with other system tools as for example hosts.deny/firewall etc...

If amavisd/spamassassin is configured the right way, no cluebringer is needed anymore.

-Paul

**Comment by carragom** [comment]
A couple of things I don't know how to do without a policy daemon are:
- Limit the number of mail a user can send per min/hour/day etc.
- Checking SPF before the email enters the system. Just like blacklist checks (which can/should be done right on postfix and not on SpamAssasin), SPF check should be done sooner rather than later. And at least in my setup, once it reaches SpamAssassin it ate like 75% of the resouces it needed.

If there is any way handle those two points without a policy daemon, please tell me how :).
Cheers.

**Comment by piwats** [comment]
True, that is something the standard setups don't do, unless you set those configuration options which limit all kind of rates already provided by postfix. However that does not allow fine-grained setting like per hour/day... (how does the use case for this looks like?)

SPF is very controversial. amavis/spammassassin is a content filter in the first place and probably should be the last checking mechanism. Still, i dont see a problem with this. About what load do we speak here?

There are so many extensions to postfix. How about a modoboa plugin, which allows to be simply extended to not only exclusively allow the configuration/provide a gui to f.e. policyd/cluebringer, but to any postfix extension desired ?

-Paul

**Comment by carragom** [comment]
Hi Paul. Thanks for your comments. I want to clarify that I did not suggested this feature. I just think it's a good idea :).

Sometimes an account can be compromise and it's here where per day/hour limit shines, preventing the attacker to use the compromised account to relay thousands of mails and instead only be able to send say two hundred per day or ten per hour. This kind of limits rarely affects regular users and can be fine tuned. I'm not aware of anything in Postfix that does anything similar to this. I know there is rate limiting but I think that plays a different role.

On the SPF case, you are right on the money, content filtering (amavis/spamassasing) is not the right place to do SPF. SPF like blacklists should be checked as early as possible to prevent the system from spending resources (CPU, Memory, IO) on an email that will be ultimately rejected. But unlike blacklists, I don't think Postfix can do this by itself. Also when the rejection happens early, it falls under the sending server (not us) responsibility to generate the DSN. When this checking is done in the content filter, since out server already accepted the email, it becomes our responsibility to generate the DSN spending even more server resources and possibility be the source of backscatter mail.

I hope that helped to clarify a bit why I think it would be a good idea to have this.
Cheers.

**Comment by patrickbenkoetter** [comment]
- carragom reply@reply.github.com:
  > Hi Paul. Thanks for your comments. I want to clarify that I did not suggested this feature. I just think it's a good idea :).
  > 
  > Sometimes an account can be compromise and it's here where per day/hour limit shines, preventing the attacker to use the compromised account to relay thousands of mails and instead only be able to send say two hundred per day or ten per hour. This kind of limits rarely affects regular users and can be fine tuned. I'm not aware of anything in Postfix that does anything similar to this. I know there is rate limiting but I think that plays a different role.

postfwd does that and IMO it does a better job at it. Besides the fact, that
it is a policy server and not a MILTER. MILTERs take you further. They begin
at client IP (session) and go all the way until the end (END OF DATA). Policy
servers are restricted to SMTP session only. In the long run a MILTER will
gain us more. I suggest we think about using a MILTER.

> On the SPF case, you are right on the money, content filtering (amavis/spamassasing) is not the right place to do SPF. SPF like blacklists should be checked as early as possible to prevent the system from spending resources (CPU, Memory, IO) on an email that will be ultimately rejected. But unlike blacklists, I don't think Postfix can do this by itself. Also when the rejection happens early, it falls under the sending server (not us) responsibility to generate the DSN. When this checking is done in the content filter, since out server already accepted the email, it becomes our responsibility to generate the DSN spending even more server resources and possibility be the source of backscatter mail.

SPF will become more important as DMARC makes its way through IETF until it
will be standardized (and fixed as it is currently br0ken in some parts).

p@rick

> I hope that helped to clarify a bit why I think it would be a good idea to have this.
> Cheers.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/tonioo/modoboa/issues/575#issuecomment-48864702

## 

[*] sys4 AG

https://sys4.de, +49 (89) 30 90 46 64
Franziskanerstraße 15, 81669 München

Sitz der Gesellschaft: München, Amtsgericht München: HRB 199263
Vorstand: Patrick Ben Koetter, Marc Schiffbauer
Aufsichtsratsvorsitzender: Florian Kirstein

**Comment by almereyda** [comment]
Is this discussion still about `policyd`, else it could be renamed or closed.
People seem to like Modoboa for using `postscreen` instead: https://twitter.com/TwizzyDizzy/status/732204315511783425

**Comment by hadifarnoud** [comment]
I was looking for policyD integration as well. My issue is some users on my server send bulk emails and that puts everyone's emails in spam. it is very hard to find the responsible account manually.

so, I understand policyD or postfwd will give me ability to limit send per account/domain.

is there a way I can set it up right now? do you have any ETA?

**Comment by tonioo** [comment]
@hadifarnoud No ETA for this but if your need is just about limiting the sending rate, you can already do it using postfix (example: https://wiki.deimos.fr/Postfix:_limit_outgoing_mail_throttling.html).

**Comment by hadifarnoud** [comment]
@tonioo I want to set limit per sending account/domain. that does the opposite.

update: and looks like there is already a transport map here `/etc/postfix/sql-transport.cf`
ps: I can't connect to database for some reason
`ERROR 1045 (28000): Access denied for user 'modoboa'@'localhost' (using password: YES)`
